### PR TITLE
Fix Scrollspy to work with all valid HTML5 element ids

### LINKED ISSUE #34381: Scrollspy does not handle all valid HTML5 element ids
## Bug Report

Scrollspy stops working when any of the DOM elements it is "spying" on has an id that starts with a number, such as

```html
<div id="3rd">Text</div>
```

This should be considered a bug because HTML5 has no restrictions on what form an ID can take ([official spec](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:the-id-attribute-2)).
> There are no other restrictions on what form an ID can take; in particular, IDs can consist of just digits, start with a digit, start with an underscore, consist of just punctuation, etc.

Historically, HTML4 did not allow for ids to start with numbers, but it would seem ridiculous to assume only HTML4 content today.

## Bug Demo

See this JSBin for the demo. https://jsbin.com/rehematazo/3/edit?html,output. By default, I've commented out the `<a>` elements that link to `<div>`s whose IDs start with a digit, so Scrollspy works. But once you uncomment out these `<a>` tags, then Scrollspy stops working entirely.

I have also copied the HTML code for the demo here:
<details>
<summary>Code</summary>

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>JS Bin</title>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
  <style>
    .tall { height: 300px; }
    
    .toc-nav { /* <nav> */
      position: sticky;
      top: 0;
      height: min-content;
      max-height: 100vh;
      overflow-y: auto;
    }
    
    .toc-nav a.active {
      color: #007bff;
    }
  </style>
</head>
<body data-bs-spy="scroll" data-bs-target="#nav">

  <div class="container"><div class="row">

  <main class="col-8">
  <h2 id="A">ID A</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="B">ID B</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="C">ID C</h2>
  <div class="tall">Chunk of text</div>

  <h2 id="1">ID 1</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="2">ID 2</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="3">ID 3</h2>
  <div class="tall">Chunk of text</div>
  </main>
  
  <nav class="col-4 toc-nav" id="nav">
    <nav class="navbar navbar-light">
      <span class="navbar-brand">Contents</span>
      <a class="list-group-item list-group-item-action" href="#A">Item A</a>
      <a class="list-group-item list-group-item-action" href="#B">Item B</a>
      <a class="list-group-item list-group-item-action" href="#C">Item C</a>

      <!-- Uncomment the following 3 lines to see the bug:
           Once you uncomment, you will notice that Scrollspy stops working.
        -->
      <!-- <a class="list-group-item list-group-item-action" href="#1">Item 1</a>
      <a class="list-group-item list-group-item-action" href="#2">Item 2</a>
      <a class="list-group-item list-group-item-action" href="#3">Item 3</a> -->
    </nav>
  </nav>

  </div></div>

  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.min.js" integrity="sha384-cVKIPhGWiC2Al4u+LWgxfKTRIcfu0JTxR+EQDz/bgldoEyl4H0zUF0QKbrJ0EcQF" crossorigin="anonymous"></script>
</body>
</html>
```

</details>

## Why this happens, and possible fixes

This bug seems to happen because Bootstrap uses [`document.querySelector()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) to identify elements:

https://github.com/twbs/bootstrap/blob/696996dd4f5b7319413655b9134ca2f3768c2549/js/dist/scrollspy.js#L78

However, `document.querySelector()` only works for valid CSS selector strings, and CSS selector strings for IDs cannot start with a number (even though HTML IDs can). See this StackOverflow Q&A for more info: https://stackoverflow.com/q/20306204. This causes Bootstrap JS to encounter the following error (copied from Microsoft Edge console):

```
Uncaught DOMException: Failed to execute 'querySelector' on 'Document': '#3rd' is not a valid selector.
    at <anonymous>:1:10
```

The StackOverflow Q&A proposes several possible fixes. For example, `document.getElementById("3rd")` works fine, but only works with ID-based selectors. I am also not sure which fixes fall within Bootstrap's browser compatibility goals.

## System Info

OS: Windows 10 20H1
Browser: Microsoft Edge (Chromium) 91.0.864.59 and Firefox 89.0.2 (64-bit)
Bootstrap version 5.0.2

### PR DESCRIPTION
**Pull request by chrisyeh96**
Fixes #34381 by using `document.getElementById()` instead of `document.querySelector()` when testing the validity of an element ID.

Whereas `document.querySelector()` only allows valid CSS selectors and requires ID-based selectors to start with a letter, `document.getElementById()` handles all valid HTML5 element IDs.

### CODE REVIEW

File: js/src/scrollspy.js [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import {
   defineJQueryPlugin,
   getElement,
-  getSelectorFromElement,
+  getTargetIdFromElement,
```
Comment (GeoSot) [inline js/src/scrollspy.js:11]: Thank you @chrisyeh96 for you contribution :smiley:  

Instead of introducing another function to just make the job, you could probably make a minor refactor to

`getElementFromSelector` to check if string starts with `#` and use the `getElementById`. 

After this you can use it inside `getSelectorFromElement` as they use the exactly same code :wink:

Comment (chrisyeh96) [reply js/src/scrollspy.js:11]: Unfortunately, I don't think that your suggestion will work with edge cases. For example `#someID > div` is a valid CSS Selector that starts with a `#` but cannot be passed directly to `document.getElementById()`.

I wrote the new function `getTargetIdFromElement()` specifically to be used in cases such as Scrollspy where the Bootstrap documentation explicitly says that the `href` or `data-bs-target` attributes of the element have the format `#someValidHTML5Id`.

The more general `getElementFromSelector()` should still be allowed for generic CSS selectors such as the example above (`#someID > div`).

Comment (GeoSot) [reply js/src/scrollspy.js:11]: > Unfortunately, I don't think that your suggestion will work with edge cases. For example #someID > div is a valid CSS Selector that starts with a # but cannot be passed directly to document.getElementById().

Maybe this could do the job. Right?

```js

reg = new RegExp(/^#[\w\-]+$/gm)
reg.test(givenId)
```

Comment (chrisyeh96) [reply js/src/scrollspy.js:11]: That doesn't work either. As I mentioned in my original issue #34381, any Unicode character that isn't a whitespace is considered a valid HTML5 ID. In particular, numbers, emoji, and non-Latin characters are all valid. These are not matched by your suggested regex.

**My pitch for a separate `getTargetIdFromElement()` method**

In my mind, there are 2 distinct cases that Bootstrap should support:

*Case 1*: Arbitrary CSS Selectors

The Bootstrap docs for components such as [Collapse](https://getbootstrap.com/docs/5.0/components/collapse/#multiple-targets) and [Offcanvas](https://getbootstrap.com/docs/5.0/components/offcanvas/#via-data-attributes) specifically mention supporting arbitrary CSS selectors. These should use the existing `getSelectorFromElement()` / `getElementFromSelector()` methods, which call `document.querySelector()` behind the scenes. By definition of CSS selectors (see official spec below), these components would NOT support IDs that start with digits.

*Case 2*: ID-specific Selectors

The Bootstrap docs for components such as [Scrollspy](https://getbootstrap.com/docs/5.0/components/scrollspy/) and [Carousel](https://getbootstrap.com/docs/5.0/components/carousel/#example) explicitly mention requiring matching an HTML5 ID (see official spec below). These should use the new `getTargetIdFromElement()` method which only matches HTML5 IDs instead of arbitrary CSS selectors. Thus, these components would be able to support IDs that start with digits.

<details><summary>Official Specs for HTML5 IDs and CSS Selectors</summary>

**HTML5 ID**: The [official spec](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:the-id-attribute-2) states
>When specified on HTML elements, the id attribute value must be unique amongst all the IDs in the element's tree and must contain at least one character. The value must not contain any ASCII whitespace. [...] There are no other restrictions on what form an ID can take; in particular, IDs can consist of just digits, start with a digit, start with an underscore, consist of just punctuation, etc.

The method `document.getElementById()` supports all valid HTML5 IDs.

**CSS Selectors**: The [official spec](https://www.w3.org/TR/selectors-4/#id-selector) states
> An ID selector consists of a “number sign” (U+0023, #) immediately followed by the ID value, which must be a CSS identifier

where a CSS identifier is defined ([official spec](https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier)) as follows:
> In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit.

The method `document.querySelector()` only supports valid CSS selectors. For example, `document.querySelector("#3rd")` will raise an exception.

</details>

File: js/src/util/index.js [hunk 2]
```diff
@@ -34,6 +34,19 @@ const getUID = prefix => {
...
 }
 
+const getTargetIdFromElement = element => {
+  const selector = element.getAttribute('data-bs-target') || element.getAttribute('href')
+
+  if (selector) {
+    // a valid HTML5 id is any string without whitespace
+    const match = selector.match(/#(\S+)/)
+    const id = match ? match[1] : null
+    return document.getElementById(id) ? id : null
```
Comment (XhmikosR) [inline js/src/util/index.js:44]: Can't this be tweaked a bit further? Like, there's no need to even call `getElementById` if `id` is `null`.

Comment (chrisyeh96) [reply js/src/util/index.js:44]: True, but `getElementById()` already performs a check against `null` anyway.

**As-is**
`getElementById()` always performs a single check on whether `id` is null.

**Your proposed tweak**
If `id` is null, then we perform a single check on whether it's null. This is probably slightly faster than the function call to `getElementById()`.
If `id` is not null, then we still perform a check ourselves, but then `getElementById()` also performs a check. In other words, there are two branching statements.

In other words, your proposal may speed up the case when `id` is null, but it's slower when `id` is not null. I'm happy with either option. Feel free to change it as you see fit.

**Comment by chrisyeh96** [comment]
@XhmikosR @GeoSot Any update on this? What more needs to be done to get this merged?
